File: library/src/main/java/me/proxer/library/api/list/IndustryProjectListEndpoint.java
package me.proxer.library.api.list;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.PagingLimitEndpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entitiy.list.IndustryProject;
import me.proxer.library.enums.IndustryType;

import javax.annotation.Nullable;
import java.util.List;

/**
 * Endpoint for retrieving a list of projects of an industry.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class IndustryProjectListEndpoint implements PagingLimitEndpoint<List<IndustryProject>> {

    private final InternalApi internalApi;

    private final String id;

    /**
     * Sets the type of industries to filter by.
     */
    @Nullable
    @Setter
    private IndustryType type;

    @Nullable
    private Integer includeHentai;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer page;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer limit;

    IndustryProjectListEndpoint(final InternalApi internalApi, final String id) {
        this.internalApi = internalApi;
        this.id = id;
    }

    @Override
    public ProxerCall<List<IndustryProject>> build() {
        return internalApi.industryProjectList(id, type, includeHentai, page, limit);
    }

    /**
     * Sets if hentai should be included in the result.
     */
    @SuppressWarnings("checkstyle:hiddenfield")
    public IndustryProjectListEndpoint includeHentai(@Nullable final Boolean includeHentai) {
        if (includeHentai == null) {
            this.includeHentai = null;
        } else {
            this.includeHentai = includeHentai ? 0 : -1;
        }

        return this;
    }
}

File: library/src/main/kotlin/me/proxer/library/api/list/IndustryProjectListEndpoint.kt
package me.proxer.library.api.list

import me.proxer.library.ProxerCall
import me.proxer.library.api.PagingLimitEndpoint
import me.proxer.library.entity.list.IndustryProject
import me.proxer.library.enums.IndustryType

/**
 * Endpoint for retrieving a list of projects of an industry.
 *
 * @author Ruben Gees
 */
class IndustryProjectListEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val id: String
) : PagingLimitEndpoint<List<IndustryProject>> {

    private var page: Int? = null
    private var limit: Int? = null

    private var type: IndustryType? = null
    private var includeHentai: Boolean? = null

    override fun page(page: Int?) = this.apply { this.page = page }
    override fun limit(limit: Int?) = this.apply { this.limit = limit }

    /**
     * Sets the type of industries to filter by.
     */
    fun type(type: IndustryType?) = this.apply { this.type = type }

    /**
     * Sets if hentai should be included in the result.
     */
    fun includeHentai(includeHentai: Boolean? = true) = this.apply { this.includeHentai = includeHentai }

    override fun build(): ProxerCall<List<IndustryProject>> {
        return internalApi.industryProjectList(id, type, includeHentai, page, limit)
    }
}
